src/EVM/Fuzz.hs

```diff
@@ -0,0 +1,238 @@
+tryCexFuzz ps tries = unGen (testVals tries) (mkQCGen 0) 1337
+  where
+    vars = extractVars ps
+    bufs = extractBufs ps
+    stores = extractStorage ps
+    testVals :: Integer -> Gen (Maybe SMTCex)
+    testVals 0 = pure Nothing
+    testVals todo = do
+      varVals <- getVals vars
+      bufVals <- getBufs bufs
+      storeVals <- getStores stores
+      let
+        ret =  filterCorrectKeccak $ map (substituteEWord varVals . substituteBuf bufVals . substituteStores storeVals) ps
+        retSimp =  Expr.simplifyProps ret
+      if null retSimp then pure $ Just (SMTCex {
+                                    vars = varVals
+                                    , addrs = mempty
+                                    , buffers = bufVals
+                                    , store = storeVals
+                                    , blockContext = mempty
+                                    , txContext = mempty
+                                    })
+                                    else testVals (todo-1)
+
+-- Filter out PEq (Lit X) (keccak (ConcreteBuf Y)) if it's correct
+filterCorrectKeccak :: [Prop] -> [Prop]
+filterCorrectKeccak ps = filter checkKecc ps
+  where
+    checkKecc (PEq (Lit x) (Keccak (ConcreteBuf y))) = keccak' y /= x
+    checkKecc _ = True
+
+substituteEWord :: Map (Expr EWord) W256 -> Prop -> Prop
+substituteEWord valMap p = mapProp go p
+  where
+    go :: Expr a -> Expr a
+    go orig@(Var _) = Lit (valMap ! orig)
+    go orig@(TxValue) = Lit (valMap ! orig)
+    go a = a
+
+
+substituteBuf :: Map (Expr Buf) BufModel -> Prop -> Prop
+substituteBuf valMap p = mapProp go p
+  where
+    go :: Expr a -> Expr a
+    go orig@(AbstractBuf _) = case (valMap !? orig) of
+                                Just (Flat x) -> ConcreteBuf x
+                                Just (Comp _) -> internalError "No compressed allowed in fuzz"
+                                Nothing -> orig
+    go a = a
+
+substituteStores ::  Map (Expr 'EAddr) (Map W256 W256) -> Prop -> Prop
+substituteStores valMap p = mapProp go p
+  where
+    go :: Expr a -> Expr a
+    go (AbstractStore a) = case valMap !? a of
+                                  Just m -> ConcreteStore m
+                                  Nothing -> ConcreteStore mempty
+    go a = a
+
+-- Var extraction
+data CollectVars = CollectVars {vars :: Set.Set (Expr EWord)
+                               ,vals :: Set.Set W256
+                               }
+  deriving (Show)
+
+initVarsState :: CollectVars
+initVarsState = CollectVars {vars = Set.empty
+                            ,vals =  Set.empty
+                            }
+
+findVarProp :: Prop -> State CollectVars Prop
+findVarProp p = mapPropM go p
+  where
+    go :: forall a. Expr a -> State CollectVars (Expr a)
+    go = \case
+      e@(Var a) -> do
+        s :: CollectVars <- get
+        put CollectVars {vars=Set.insert (Var a) s.vars, vals = s.vals}
+        pure e
+      e@(Lit a) -> do
```

Comment:
Why are literals needed here?

Reply:
What we do here is that we extract known literals and then later, we sometimes set values of variables to these known literals:

```
      pickKnown :: Bool <- arbitrary
      val <- if (not pickKnown) || (null vars.vals) then do getRndW256
                                                    else elements $ Set.toList (vars.vals)
```

So if there is a check like `if (a == 57) then ...` we will pick out `57` as a `Lit` and later try to set variables, e.g. `a` to 57. Basically, we are taking out magic constants from the `Expr`. Magic constant extraction is a known technique in this space. Here's Echidna fuzzer doing it:

https://github.com/crytic/echidna/blob/f52b35af4e646da3d9d86408d9ccda9134e2b9c7/lib/Echidna.hs#L74

and:

https://github.com/crytic/echidna/blob/f52b35af4e646da3d9d86408d9ccda9134e2b9c7/lib/Echidna/Processor.hs#L56